### PR TITLE
support `.closure` iterators with all backends

### PR DESCRIPTION
**Pull request by zerbina**
## Summary

`jsgen` and `vmgen` now implement the `finished` magic procedure, making `.closure` iterators available:
- at compile time
- for the NimScript target
- for the standalone VM target
- for the JavaScript target

## Details

The `finished` magic procedure is responsible for querying wether the provided iterator has finished, and is inserted by the compiler when transforming for-loops that use `.closure` iterators.

Implementing support for it in the JavaScript code-generator is straightforward, as the operation can be mapped to a simple field lookup + comparison.

Closures (which includes `.closure` iterator instances) are not represented as normal tuples in the VM, but rather via a dedicated type that uses dedicated operations for querying and modification. To be able to access the environment of a closure object, the new ``AccessEnv`` operation is introduced, which the `mFinished` implementation in `vmgen` then makes use of.

The compiler run-time procedure `closureIterExcSetup` is also implemented for both the JavaScript and VM target, which required the VM callback handling slightly adjusted in order to support setting the virtual machine's current exception.

### Tests

- enable the tests that make use of `.closure` iterators for the JavaScript and VM target
- reduce the test for a closure bug, removing the unnecessary dependency on the `os` module. Since the bug is unrelated to `.closure` iterators, it is moved to a different test file
- move a test unrelated to `hash` out of `thashes.nim`
- make a test in `ttemplate_issues.nim` work when located inside a `block`

---

Thank you again for preparing the `lang_callable` category, @saem.

## Notes for Reviewers
* this is a follow-up to #607 ; most of the work towards full `.closure` iterator support happened there

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

### CODE REVIEW

File: tests/lang_callable/closure/tclosure_nested_iterator.nim [hunk 1]
```diff
@@ -1,5 +1,5 @@
 discard """
-  targets: "c !js !vm"
+  targets: "c js vm"
```
Comment (saem) [inline tests/lang_callable/closure/tclosure_nested_iterator.nim]: Could delete this line outright.

**Review by saem — APPROVED** [review]
Awesome possum.

**Comment by saem** [comment]
I'll kick off bors, worst case it can be cancelled, the incredibly minor suggestion can be tackled anytime someone sees it next.

bors r+

**Comment by zerbina** [comment]
Ah, sorry. I missed the comment by one minute :D.

**Comment by saem** [comment]
I laughed when I saw the push notification. 😆 

bors r+

**Pull request merged**